### PR TITLE
Feature/UI-16 Smart List Search

### PR DESCRIPTION
**Pull request by lmiller1990**
~~TODO: I'll add a test soon. But the implementation is done.~~ Added a test. Abandoning enzyme because it sucks, react testing library is much better and closer to how a user interacts with the app.

image

### CODE REVIEW

File: src/SmartList/components/SearchInput/SearchInput.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+
```
Comment (MichaelDorkhom) [inline src/SmartList/components/SearchInput/SearchInput.tsx]: Remove blank line

File: src/SmartList/components/SearchInput/SearchInput.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import React, { HTMLProps } from 'react'
+import Form from 'react-bootstrap/Form'
+import InputGroup from 'react-bootstrap/InputGroup'
+import { ReplaceProps, BsPrefixProps } from 'react-bootstrap/helpers'
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { faSearch } from '@fortawesome/free-solid-svg-icons'
+
+import { FormControlProps } from 'react-bootstrap'
```
Comment (MichaelDorkhom) [inline src/SmartList/components/SearchInput/SearchInput.tsx]: No blank line above this

File: src/SmartList/components/SearchInput/SearchInput.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+export type TExtendsHTMLElement<T> = Omit<HTMLProps<HTMLElement>, keyof T> & T
+
+interface IProps {
+  value: string
+  onChange: (value: string) => any
+}
+
+const SearchInput: React.FC<TExtendsHTMLElement<IProps>> = (props): JSX.Element => {
+  const handleChange = (e: React.FormEvent<ReplaceProps<'input', BsPrefixProps<'input'> & FormControlProps>>) => {
+    const value = e.currentTarget && e.currentTarget.value && e.currentTarget.value
```
Comment (MichaelDorkhom) [inline src/SmartList/components/SearchInput/SearchInput.tsx]: `e.currentTarget.value` checked twice

File: src/SmartList/components/SearchInput/SearchInput.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    props.onChange(value || '')
+  }
+
+  return (
+    <div className='search-wrapper'>
+      <InputGroup>
+      <Form.Control
```
Comment (MichaelDorkhom) [inline src/SmartList/components/SearchInput/SearchInput.tsx]: Indentation is not right here

File: src/SmartList/components/SearchInput/index.scss [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  .smart-list-search {
+    border-right: none;
+  }
+
+  .search-icon-wrapper.input-group-text {
+    width: 40px;
+    background: white;
+    border-left: none;
+  }
+}
```
Comment (MichaelDorkhom) [inline src/SmartList/components/SearchInput/index.scss]: next line

File: src/SmartList/components/SearchInput/SearchInput.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import Form from 'react-bootstrap/Form'
+import InputGroup from 'react-bootstrap/InputGroup'
+import { ReplaceProps, BsPrefixProps } from 'react-bootstrap/helpers'
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { faSearch } from '@fortawesome/free-solid-svg-icons'
+
+import { FormControlProps } from 'react-bootstrap'
+import './index.scss'
+
+export type TExtendsHTMLElement<T> = Omit<HTMLProps<HTMLElement>, keyof T> & T
```
Comment (MichaelDorkhom) [inline src/SmartList/components/SearchInput/SearchInput.tsx]: Don't need this, can import from `src/types.ts`

File: src/SmartList/SmartList.tsx [hunk 7]
```diff
@@ -25,24 +27,55 @@ export interface IRowElement {
...
+      <SearchInput
+        onChange={setSearchText}
+        value={searchText}
+      />
+    </div>
   )
 
   return (
-    <div>
+    <div className=''>
```
Comment (MichaelDorkhom) [inline src/SmartList/SmartList.tsx]: Don't need className here

File: src/SmartList/SmartList.tsx [hunk 8]
```diff
@@ -25,24 +27,55 @@ export interface IRowElement {
...
   )
 
-  const listContent = data.map(row =>
-    <ListItem
-      onClick={row.onClick}
-      key={row.id}
-      columns={row.columns}
-    />
+  const listContent = data.reduce<JSX.Element[]>((acc, row) => {
+    if (!searchText.length) {
```
Comment (MichaelDorkhom) [inline src/SmartList/SmartList.tsx]: Can you merge the two if statements since they're returning the same thing

Comment (lmiller1990) [reply src/SmartList/SmartList.tsx]: Updated this reduce call, please review

File: src/SmartList/components/SearchInput/SearchInput.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+              icon={faSearch}
+            />
+          </InputGroup.Text>
+        </InputGroup.Append>
+      </InputGroup>
+    </div>
+  )
+}
+
+export { SearchInput }
```
Comment (MichaelDorkhom) [inline src/SmartList/components/SearchInput/SearchInput.tsx]: Next line

File: src/SmartList/SmartList.tsx [hunk 10]
```diff
@@ -25,29 +27,66 @@ export interface IRowElement {
...
+          key={row.id}
+          columns={row.columns}
+        />
+      )
+      return acc
+    }
+
+    // if there is some search text, we stringify the content and see if
+    // any of them match the search string
+    if (
```
Comment (MichaelDorkhom) [inline src/SmartList/SmartList.tsx]: Can you merge the two conditions

File: src/SmartList/components/SearchInput/SearchInput.tsx (hunk 9 above)
Comment (lmiller1990) [reply src/SmartList/components/SearchInput/SearchInput.tsx]: why do people add new lines to the end of a file? It doesn't seem to serve any purpose

**Pull request merged**